### ISSUE TITLE
Caching mirrors

### ISSUE THREAD

**Issue by brianjmurrell** [fedora-infra/mirrormanager2#304]
There are tools that allow one to _mirror on demand_ as a caching proxy mirror.  Such mirrors of course are _virtually_ complete but in reality only as complete as the clients that use it have requested files/packages from it.

The problem with this is that `report_mirror` may report only a partial listing, but the reality is that the mirror is _complete_ from the perspective of it's clients.

It would be useful if mirrormanager and/or `report_mirror` could be informed of this so that partial `report_mirror`s don't make invalidate the mirror.  In the short-term `report_mirror` could be modified to _fake_ the output of a full mirror, for example.

**Comment by adrianreber** [comment]
There is the flag `Always up to date`. We use it in our Fedora and CentOS Stream setups to mark primary mirrors and cloudfront based caching setups. 

![image](https://user-images.githubusercontent.com/8724298/146796933-42c05350-0cb4-43a3-be4a-40c0e08d353e.png)

Does that help?

**Comment by brianjmurrell** [comment]
That might, if I could find where it's set.  I've looked in both my site and host configuration and don't see that option.

**Comment by adrianreber** [comment]
> That might, if I could find where it's set. I've looked in both my site and host configuration and don't see that option.

It is an admin only option.

**Comment by brianjmurrell** [comment]
Admin for the [private] mirror?  I should be admin for my own private mirror.  I can change many other aspects of my site and host.

**Comment by adrianreber** [comment]
Ah, private mirror. Right. No, it is a mirrormanager instance admin only option. For private mirrors you cannot set it yourself.

It doesn't sound problematic for private mirrors but currently it is not implemented.

**Comment by brianjmurrell** [comment]
So, back to [original problem description](https://github.com/fedora-infra/mirrormanager2/issues/304#issue-1084950590).  :-)

**Comment by adrianreber** [comment]
You could fake it by creating the directory tree with empty directories. `report_mirror` only scans directories (if I remember it correctly) to reduce I/O load. In theory a directory tree with no files should lead to an up to date mirror via `report_mirror`

**Comment by brianjmurrell** [comment]
Indeed.  But what should all be in such an empty directory tree?  And it does change from release to release (i.e. the introduction of `Modular` for example) so what would need to be created would need to be different on a per release basis.

This issue is all about resolving this problem given those constraints.

**Comment by brianjmurrell** [comment]
I suppose one could do something like:

```
# rsync -nai -f"+ */" -f"- *" rsync://<mirror>/fedora/releases/35 /path/to/local/mirror/fedora/releases/
```

To mimic the directory structure.  Just sucks that this needs to be such a manual operation on every new release of a distro.

**Comment by brianjmurrell** [comment]
I'm not sure what you are implying or trying to say @ott but I can tell you, from lots of experience -- both professional (i.e.`$dayjob` where we do a shedload of RPM mirroring -- think CI systems that get re-installed for every CI run) and personal, that HTTP caching Fedora mirrors works quite well.

**Comment by brianjmurrell** [comment]
@ott I don't doubt your experience.  But I also don't doubt my literally decades of experience with running RPM/DEB mirror caching proxies with great success.  You can continue to tell me that it won't work, but again, I have decades of experience with them telling me it does work.

To be perfectly clear, I am not referring to using generic HTTP proxies like [Squid Cache](https://wiki.squid-cache.org/SquidFaq/AboutSquid) but purpose-built RPM/DEB (and potentially more package formats) proxy caches like [Nexus Repository Manager](https://www.sonatype.com/products/sonatype-nexus-repository#), [Artifactory](https://jfrog.com/artifactory/) and even the much more lightweight [pkg-cacher](https://github.com/open-sw/pkg-cacher) and/or [AptProxy](https://help.ubuntu.com/community/AptProxy).

**Comment by brianjmurrell** [comment]
> As I said, specialized HTTP caches that can use repository metadata might work but you did not mention such software in your statement.

I didn't need to.  I didn't come here to report problems with my caching mirror software so it was not relevant.  The discussion of such is only because you started positing that it doesn't/won't/can't work, quite OT for this issue I might add.

> Moreover, software that is written in an interpreted language and that has to serve every request, might not be suitable to high-volume mirrors.

I never said I was operating a high-volume mirror with an interpreted language solution .  Quite the opposite in fact as the solution that I operate with an interpreted language tool is a private mirror for a small network, so it works sufficiently well.

While I appreciate your perspective on proxy/caching RPM repos, given that I am doing what I am doing with proxy/caching repos quite successfully and have for many many years, you are not going to convince me that it can't work and that I should stop doing it.
